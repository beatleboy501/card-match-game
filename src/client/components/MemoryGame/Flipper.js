import React from 'react';
import PropTypes from 'prop-types';

class Flipper extends React.Component {
  constructor(props) {
    super(props);
    const { isClosed } = this.props;
    this.state = {
      isClosed,
      rotation: 0,
    };
  }

  componentWillReceiveProps(nextProps) {
    const { isClosed } = this.props;
    const { rotation } = this.state;
    if (nextProps.isClosed !== isClosed) {
      this.setState({ isClosed: nextProps.isClosed });
      this.setState({ rotation: rotation + 180 });
    }
  }

  getComponent(key) {
    const { children } = this.props;
    return children.filter(component => component.key === key);
  }

  render() {
    const { isClosed } = this.state;
    const styles = {
      container: {
        perspective: '1000px',
        transformStyle: 'preserve-3d',
        margin: '50px',
      },
      flipper: {
        position: 'relative',
        transform: 'rotateY(0deg)',
        transformStyle: 'preserve-3d',
        transition: '0.6s',
      },
      flipperFlip: {
        position: 'relative',
        transform: 'rotateY(180deg)',
        transformStyle: 'preserve-3d',
        transition: '0.6s',
      },
      front: {
        WebkitBackfaceVisibility: 'hidden',
        backfaceVisibility: 'hidden',
        left: '0',
        position: 'absolute',
        top: '0',
        transform: 'rotateY(0deg)',
        transformStyle: 'preserve-3d',
        width: '100%',
        zIndex: '2',
      },
      back: {
        WebkitBackfaceVisibility: 'hidden',
        backfaceVisibility: 'hidden',
        left: '0',
        position: 'absolute',
        transform: 'rotateY(180deg)',
        transformStyle: 'preserve-3d',
        top: '0',
        width: '100%',
      },
    };

    return (
      <div className="Flipper" style={styles.container}>
        <div style={isClosed ? styles.flipperFlip : styles.flipper}>
          <div style={styles.front}>
            {this.getComponent('front')}
          </div>
          <div style={styles.back}>
            {this.getComponent('back')}
          </div>
        </div>
      </div>
    );
  }
}

Flipper.propTypes = {
  cardStyles: PropTypes.shape({
    front: PropTypes.object,
    back: PropTypes.object,
  }),
  children: (props, propName, componentName) => {
    if (React.Children.count(props[propName]) !== 2) {
      return new Error(`${componentName} requires two children.`);
    }
    return null;
  },
  isClosed: PropTypes.bool,
};

Flipper.defaultProps = {
  cardStyles: {
    front: {},
    back: {},
  },
  children: [],
  isClosed: true,
};

export default Flipper;
