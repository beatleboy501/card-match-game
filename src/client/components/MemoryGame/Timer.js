import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  resetTimer, setDiff, startTimer, stopTimer, tick,
} from '../../redux/actions/timerActions';
import { addZero } from '../../utils';

const propTypes = {
  diff: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.number]),
  interval: PropTypes.number,
  startTime: PropTypes.number,
  dispatchStartTimer: PropTypes.func.isRequired,
  dispatchStopTimer: PropTypes.func.isRequired,
  dispatchResetTimer: PropTypes.func.isRequired,
  dispatchSetDiff: PropTypes.func.isRequired,
  dispatchTick: PropTypes.func.isRequired,
};

const defaultProps = {
  startTime: null,
  diff: null,
  interval: null,
};

export class Timer extends React.Component {
  constructor(props) {
    super(props);
    this.getTime = this.getTime.bind(this);
    this.start = this.start.bind(this);
    this.stop = this.stop.bind(this);
    this.reset = this.reset.bind(this);
    this.tick = this.tick.bind(this);
    this.getDiff = this.getDiff.bind(this);
    this.setDiff = this.setDiff.bind(this);
  }

  getDiff() {
    const { diff } = this.props;
    return diff;
  }

  setDiff(diff) {
    const { dispatchSetDiff } = this.props;
    dispatchSetDiff(diff);
  }

  getTime() {
    const { diff } = this.props;
    let hundredths = diff ? Math.round(diff.getMilliseconds() / 10) : 0;
    const seconds = diff ? diff.getSeconds() : 0;
    const minutes = diff ? diff.getMinutes() : 0;
    if (hundredths === 100) hundredths = 0;
    return { hundredths, seconds, minutes };
  }

  start() {
    const { dispatchStartTimer, startTime, interval } = this.props;
    if (interval) cancelAnimationFrame(interval);
    const newStartTime = startTime || +new Date();
    dispatchStartTimer(newStartTime, requestAnimationFrame(this.tick));
  }

  stop() {
    const { dispatchStopTimer, interval } = this.props;
    cancelAnimationFrame(interval);
    dispatchStopTimer();
  }

  reset() {
    const { dispatchResetTimer, interval } = this.props;
    cancelAnimationFrame(interval);
    dispatchResetTimer();
  }

  tick() {
    const { dispatchTick, startTime } = this.props;
    dispatchTick(new Date(+new Date() - startTime), requestAnimationFrame(this.tick));
  }

  handleGameComplete(matchesCount) {
    this.stop();
    const time = this.getTime();
    alert(`${matchesCount} matches in ${addZero(time.minutes)}:${addZero(time.seconds)}:${addZero(time.hundredths)}`);
  }

  render() {
    const { hundredths, seconds, minutes } = this.getTime();
    return (
      <section>
        <h1>
          {addZero(minutes)}
:
          {addZero(seconds)}
:
          {addZero(hundredths)}
        </h1>
      </section>
    );
  }
}

Timer.propTypes = propTypes;
Timer.defaultProps = defaultProps;

function mapStateToProps(state) {
  return { ...state.timer };
}

const mapDispatchToProps = dispatch => ({
  dispatchStartTimer: (startTime, interval) => {
    dispatch(startTimer(startTime, interval));
  },
  dispatchStopTimer: (startTime) => {
    dispatch(stopTimer(startTime));
  },
  dispatchResetTimer: () => {
    dispatch(resetTimer());
  },
  dispatchSetDiff: (diff) => {
    dispatch(setDiff(diff));
  },
  dispatchTick: (diff, interval) => {
    dispatch(tick(diff, interval));
  },
});

export default connect(mapStateToProps, mapDispatchToProps, null, { withRef: true })(Timer);
