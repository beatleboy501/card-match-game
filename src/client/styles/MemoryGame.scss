.container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  flex-direction: column;
}

.cardsContainer {
  margin: 0 auto;
  max-width: 1200px;
  padding: 0 1rem;
}

.cards {
  justify-content: flex-start;
  align-content: stretch;
  padding: 0;
  max-width: 100%;
}

.matches {
  margin-top: 1em !important;
}
