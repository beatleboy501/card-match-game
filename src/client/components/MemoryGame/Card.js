import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Flipper from './Flipper';
import sassStyles from '../../styles/Card.scss';
import logo from '../../../public/timesLogo.png';

const propTypes = {
  cardInfo: PropTypes.object.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
};

const defaultProps = {};

const styles = {
  card: {
    width: 100,
    height: 175,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    marginBottom: 16,
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
  cardBack: {
    height: 50,
    width: 50,
  },
};

export const GameCard = (props) => {
  const { cardInfo, isOpen, onClick } = props;
  return (
    <div className={sassStyles.responsiveCell}>
      <Flipper isClosed={!isOpen}>
        <Card onClick={onClick} style={styles.card} key="front">
          <CardContent>
            <div>
              <Typography color="textSecondary">
                {' '}
              </Typography>
              <Typography
                variant="headline"
                component="h2"
              >
                {cardInfo.value}
              </Typography>
            </div>
          </CardContent>
        </Card>
        <Card onClick={onClick} style={styles.card} key="back">
          <CardContent>
            <CardMedia style={styles.cardBack} image={logo} />
          </CardContent>
        </Card>
      </Flipper>
    </div>
  );
};

GameCard.propTypes = propTypes;
GameCard.defaultProps = defaultProps;

export default withStyles(styles)(GameCard);
