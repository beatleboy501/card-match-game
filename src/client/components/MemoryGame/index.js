import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Typography from '@material-ui/core/Typography';
import AwesomeDebouncePromise from 'awesome-debounce-promise';
import Card from './Card';
import styles from '../../styles/MemoryGame.scss';
import {
  addMatch, addOpenId, setOpenIds, setMatchedCards, startGame,
} from '../../redux/actions/memoryGameActions';
import CardMatchCounts from '../../constants/CardMatchCounts';

const propTypes = {
  cards: PropTypes.arrayOf(PropTypes.object),
  isFirstCard: PropTypes.bool,
  dispatchStartGame: PropTypes.func.isRequired,
  dispatchAddOpenId: PropTypes.func.isRequired,
  dispatchAddMatch: PropTypes.func.isRequired,
  dispatchSetOpenIds: PropTypes.func.isRequired,
  onGameStart: PropTypes.func.isRequired,
  onGameComplete: PropTypes.func.isRequired,
  openIds: PropTypes.arrayOf(PropTypes.string),
  matchedCards: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.string)),
  cardMatchCount: PropTypes.number.isRequired,
};

const defaultProps = {
  cards: [],
  openIds: [],
  matchedCards: [],
  isFirstCard: true,
};

export class MemoryGame extends React.Component {
  constructor(props) {
    super(props);
    this.handleCardClick = AwesomeDebouncePromise(this.handleCardClick.bind(this), 500);
    this.handleCardClickResult = this.handleCardClickResult.bind(this);
    this.isMatched = this.isMatched.bind(this);
    this.isGameComplete = this.isGameComplete.bind(this);
    this.completeMatch = this.completeMatch.bind(this);
  }

  isGameComplete() {
    const { cards, matchedCards, cardMatchCount } = this.props;
    return (matchedCards.length && cards.length)
      && matchedCards.length >= cards.length / cardMatchCount;
  }

  isMatched(cardId) {
    const { matchedCards } = this.props;
    const matches = matchedCards.reduce((acc, match) => acc.concat(match), []);
    return matches.includes(cardId);
  }

  handleCardClick(clickedId) {
    const {
      openIds, isFirstCard, onGameStart, dispatchStartGame, cardMatchCount, dispatchAddOpenId,
    } = this.props;
    if (isFirstCard) {
      onGameStart();
      dispatchStartGame();
    }
    if (this.isMatched(clickedId)
      || openIds.includes(clickedId)
      || openIds.length >= cardMatchCount) return;
    if (openIds.length > 0) { // One card has already been flipped
      dispatchAddOpenId(clickedId);
      setTimeout(() => {
        this.handleCardClickResult(clickedId);
      }, 1000);
    } else { // No cards have been flipped
      dispatchAddOpenId(clickedId);
    }
  }

  handleCardClickResult(clickedId) {
    const {
      cards, dispatchSetOpenIds, openIds, matchedCards, cardMatchCount,
    } = this.props;
    if (cardMatchCount === CardMatchCounts.TWO_CARD_MATCH) {
      const clickedCard = cards.find(card => card.id === clickedId);
      const openId = openIds.find(id => id !== clickedId);
      const openCard = cards.find(card => card.id === openId);
      if ((openCard && clickedCard) && openCard.value === clickedCard.value) { // there's a match
        this.completeMatch([clickedId, openId], matchedCards);
      } else { // no match
        dispatchSetOpenIds([]);
      }
    } else if (cardMatchCount === CardMatchCounts.THREE_CARD_MATCH) {
      const openCards = cards.filter(card => openIds.includes(card.id));
      const openValues = openCards.map(card => card.value);
      const distinctValues = openValues.reduce((acc, value) => {
        if (!acc.includes(value)) acc.push(value);
        return acc;
      }, []);
      if (distinctValues.length === 1) { // there's a match
        if (openValues.length === cardMatchCount) {
          this.completeMatch(openIds, matchedCards);
        }
      } else { // no match
        dispatchSetOpenIds([]);
      }
    }
  }

  completeMatch(matchArray, matchedCards) {
    const { dispatchAddMatch, dispatchSetOpenIds, onGameComplete } = this.props;
    dispatchAddMatch(matchArray);
    dispatchSetOpenIds([]);
    if (this.isGameComplete()) {
      onGameComplete(matchedCards.length);
    }
  }

  render() {
    const { cards, matchedCards, openIds } = this.props;
    return (
      <div className={styles.container}>
        <Typography className={styles.matches} variant="title" color="inherit">
          Matches
          {' '}
          {' '}
:
          {' '}
          {matchedCards.length}
        </Typography>
        <div className={styles.cardsContainer}>
          <div className={styles.cards}>
            {cards.map(card => (
              <Card
                key={card.id}
                cardInfo={card}
                isOpen={// is part of a match or it has been clicked
                  this.isMatched(card.id) || openIds.includes(card.id)
                }
                onClick={() => this.handleCardClick(card.id)}
              />
            ))}
          </div>
        </div>
      </div>
    );
  }
}

MemoryGame.propTypes = propTypes;
MemoryGame.defaultProps = defaultProps;

function mapStateToProps(state) {
  return { ...state, ...state.memoryGame };
}

const mapDispatchToProps = dispatch => ({
  dispatchAddMatch: (match) => {
    dispatch(addMatch(match));
  },
  dispatchAddOpenId: (openId) => {
    dispatch(addOpenId(openId));
  },
  dispatchSetOpenIds: (openIds) => {
    dispatch(setOpenIds(openIds));
  },
  dispatchSetMatchedCards: (matchedCards) => {
    dispatch(setMatchedCards(matchedCards));
  },
  dispatchStartGame: () => {
    dispatch(startGame());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(MemoryGame);
