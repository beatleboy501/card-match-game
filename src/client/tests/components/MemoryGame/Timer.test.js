// Use `npm run test` to run the test suite on demand;
// `npm run test-watch` will re-run it automatically with every code change.
import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Timer } from '../../../components/MemoryGame/Timer';
import { resetProps } from '../../testUtils';

Enzyme.configure({ adapter: new Adapter() });
const testInterval = 1;
const mockStartTimer = jest.fn((startTime, interval) => [startTime, interval]);
const mockStopTimer = jest.fn(() => {});
const mockResetTimer = jest.fn(() => {});
const mockSetDiff = jest.fn(newDiff => newDiff);
const mockTick = jest.fn(() => {});

let defaultProps;

let component;
let spy;
let expectedProps;

const addEventListener = jest.fn();
const requestAnimationFrame = jest.fn(() => testInterval);
const cancelAnimationFrame = jest.fn();
const removeEventListener = jest.fn();

global.window = global;
window.addEventListener = addEventListener;
window.requestAnimationFrame = requestAnimationFrame;
window.cancelAnimationFrame = cancelAnimationFrame;
window.removeEventListener = removeEventListener;

beforeEach(() => {
  defaultProps = {
    diff: new Date(0),
    startTime: Date.now(),
    interval: testInterval,
    dispatchStartTimer: mockStartTimer,
    dispatchStopTimer: mockStopTimer,
    dispatchResetTimer: mockResetTimer,
    dispatchSetDiff: mockSetDiff,
    dispatchTick: mockTick,
  };
});

afterEach(() => {
  if (spy) spy.mockClear();
  defaultProps = resetProps(defaultProps);
});

it('Renders without crashing', () => {
  component = shallow(<Timer {...defaultProps} />).shallow();
  expect(component).toMatchSnapshot();
});

test('start sets the correct props', () => {
  expectedProps = {
    diff: new Date(0),
    dispatchResetTimer: mockResetTimer,
    dispatchSetDiff: mockSetDiff,
    dispatchStartTimer: mockStartTimer,
    dispatchStopTimer: mockStopTimer,
    dispatchTick: mockTick,
    interval: testInterval,
  };
  component = mount(<Timer {...defaultProps} />);
  component.instance().start();
  expect(
    component.instance().props,
  ).toEqual(expect.objectContaining(expectedProps));
});

test('start calls correct functions', () => {
  spy = jest.spyOn(Timer.prototype, 'start');
  component = mount(<Timer {...defaultProps} />);
  component.instance().start();
  expect(spy).toHaveBeenCalled();
  expect(requestAnimationFrame).toHaveBeenCalled();
});

test('stop sets the correct props', () => {
  expectedProps = {
    diff: new Date(0),
    dispatchResetTimer: mockResetTimer,
    dispatchSetDiff: mockSetDiff,
    dispatchStartTimer: mockStartTimer,
    dispatchStopTimer: mockStopTimer,
    dispatchTick: mockTick,
    interval: testInterval,
  };
  component = mount(<Timer {...defaultProps} />);
  component.instance().stop();
  expect(
    component.instance().props,
  ).toEqual(expect.objectContaining(expectedProps));
});

test('stop calls correct functions', () => {
  spy = jest.spyOn(Timer.prototype, 'stop');
  component = mount(<Timer {...defaultProps} />);
  component.instance().stop();
  expect(spy).toHaveBeenCalled();
  expect(cancelAnimationFrame).toHaveBeenCalled();
});

test('reset sets the correct props', () => {
  expectedProps = {
    diff: new Date(0),
    dispatchResetTimer: mockResetTimer,
    dispatchSetDiff: mockSetDiff,
    dispatchStartTimer: mockStartTimer,
    dispatchStopTimer: mockStopTimer,
    dispatchTick: mockTick,
    interval: testInterval,
  };
  component = mount(<Timer {...defaultProps} />);
  component.instance().reset();
  expect(
    component.instance().props,
  ).toEqual(expect.objectContaining(expectedProps));
});

test('reset calls correct functions', () => {
  spy = jest.spyOn(Timer.prototype, 'reset');
  component = mount(<Timer {...defaultProps} />);
  component.instance().reset();
  expect(spy).toHaveBeenCalled();
  expect(cancelAnimationFrame).toHaveBeenCalled();
});
