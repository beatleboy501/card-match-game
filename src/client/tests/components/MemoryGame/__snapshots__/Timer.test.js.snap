// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <section>
    <h1>
      00
      :
      00
      :
      00
    </h1>
  </section>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1>
        00
        :
        00
        :
        00
      </h1>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "00",
          ":",
          "00",
          ":",
          "00",
        ],
      },
      "ref": null,
      "rendered": Array [
        "00",
        ":",
        "00",
        ":",
        "00",
      ],
      "type": "h1",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1>
          00
          :
          00
          :
          00
        </h1>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "00",
            ":",
            "00",
            ":",
            "00",
          ],
        },
        "ref": null,
        "rendered": Array [
          "00",
          ":",
          "00",
          ":",
          "00",
        ],
        "type": "h1",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
